### PR TITLE
fix: fix daily recurring charging profiles handling

### PR DESCRIPTION
**Pull request by jerome-benoit**

### CODE REVIEW

File: src/charging-station/ChargingStationUtils.ts [hunk 1]
```diff
@@ -671,34 +681,30 @@ const getLimitFromChargingProfiles = (
...
-      // Check if the start of the schedule is yesterday
-      if (isAfter(chargingSchedule.startSchedule, currentDate)) {
-        chargingSchedule.startSchedule.setDate(currentDate.getDate() - 1);
+      chargingSchedule.startSchedule = convertToDate(chargingSchedule.startSchedule)!;
+    }
+    // Adjust recurring start schedule
+    if (chargingProfile.chargingProfileKind === ChargingProfileKindType.RECURRING) {
+      switch (chargingProfile.recurrencyKind) {
+        case RecurrencyKindType.DAILY:
+          if (isYesterday(chargingSchedule.startSchedule)) {
```
Comment (OliveGerste) [inline src/charging-station/ChargingStationUtils.ts]: How does the code handle profiles with start schedules further in the past than yesterday?

**Review by OliveGerste — COMMENTED** [review]
I am not sure if these changes will work. 
The problem in general is that a start date can lay far in the past. 
The former algorithm did the following:
```
if (
      chargingProfile.chargingProfileKind === ChargingProfileKindType.RECURRING &&
      chargingProfile.recurrencyKind === RecurrencyKindType.DAILY &&
      isAfter(currentDate, chargingSchedule.startSchedule!)
```
checked wether it is a new profile or if it was already defined before now.
(If it is a new profile and the start lays in the future no adjustment is needed, because no limit applies in this case)
```
else if (isAfter(chargingSchedule.startSchedule!, currentDate)) {
      return null;
    }
```

If the profile is old but on a daily basis the date (not time) is adjusted to today with:
```
chargingSchedule.startSchedule.setFullYear(
        currentDate.getFullYear(),
        currentDate.getMonth(),
        currentDate.getDate(),
```

Now it is checked if the new start time isAfter the current time (this would mean the profiles virtual start date was yesterday). If the time is in the future, this means the profile started yesterday and we need to calculate the current schedule from this point of time.

**Comment by jerome-benoit** [comment]
> I am not sure if these changes will work. The problem in general is that a start date can lay far in the past. The former algorithm did the following:
> 
> ```
> if (
>       chargingProfile.chargingProfileKind === ChargingProfileKindType.RECURRING &&
>       chargingProfile.recurrencyKind === RecurrencyKindType.DAILY &&
>       isAfter(currentDate, chargingSchedule.startSchedule!)
> ```
> 
> checked wether it is a new profile or if it was already defined before now. (If it is a new profile and the start lays in the future no adjustment is needed, because no limit applies in this case)

It's not necessary to check if the profile is new or not. The only criterion for the start date adjustment is if the start date is before the current start of day or week.
If yes, the start date needs to be translate to the current day or week. The current code is buggy regarding the translation.  

> 
> ```
> else if (isAfter(chargingSchedule.startSchedule!, currentDate)) {
>       return null;
>     }
> ```

That code was incorrect, any profiles starting after now were discarded. 

> If the profile is old but on a daily basis the date (not time) is adjusted to today with:
> 
> ```
> chargingSchedule.startSchedule.setFullYear(
>         currentDate.getFullYear(),
>         currentDate.getMonth(),
>         currentDate.getDate(),
> ```
> 
> Now it is checked if the new start time isAfter the current time (this would mean the profiles virtual start date was yesterday). If the time is in the future, this means the profile started yesterday and we need to calculate the current schedule from this point of time.

Right, I will fix it the start date detection before the current recurrency window and the translation to the current window.

**Comment by OliveGerste** [comment]
>If yes, the start date needs to be translate to the current day or week. The current code is buggy regarding the translation.

Current day of week will not always work.
If it is Monday and the schedule start is Tuesday, it needs to be set to Tuesday of last week. 
Otherwise start date is in the future.

> Right, I will fix it the start date detection before the current recurrency window and the translation to the current window.
You need to be sure that the start time of the current window is never in the future, otherwise the current limit cannot be determined with the schedulePeriods

**Comment by jerome-benoit** [comment]
> > If yes, the start date needs to be translate to the current day or week. The current code is buggy regarding the translation.
> 
> Current day of week will not always work. If it is Monday and the schedule start is Tuesday, it needs to be set to Tuesday of last week. Otherwise start date is in the future.

I've used the date-fns helper to calculate the number of days or weeks to add the date in the past to move it into the current recurrency time windows. 

> 
> > Right, I will fix it the start date detection before the current recurrency window and the translation to the current window.
> 
> You need to be sure that the start time of the current window is never in the future, otherwise the current limit cannot be determined with the schedulePeriods

It will test it but it should be semantically correct if the helpers behave correctly, meaning it cannot land in the future because the difference computation in days or weeks is buggy and 'on by one'. 

The other solution is to keep the previous code setting the date to current year, month and day and checking for out of the current time windows. But it's harder to understand than using the date-fns helpers. And I'm having hard time at generalizing it for the weekly use case, the test isAfter and one day removal is only relevant for daily time windows.

**Comment by jerome-benoit** [comment]
> For instance Start Schedule: 20.July 15:00 Current time: 22.July 11:00 Difference in full days (20.July 15:00 to 22.July 23:59:59.999 ) --> 2 days. 20.July 15:00 + 2 days. --> 22.July 15:00 (date in future)

I fail to understand the issue here: if the start schedule date is at 3PM each day, it's fine to have it at the current day at also 3PM. It's just that the profile is not active yet for today time (11AM). The translation seems correct. 

I've added sanity checks in the code.

**Comment by OliveGerste** [comment]
> It's just that the profile is not active yet for today time (11AM). The translation seems correct.

Because the profile is still active from yesterday. How do you solve this? 
If the schedule is 24h this will not work.

**Comment by jerome-benoit** [comment]
> > It's just that the profile is not active yet for today time (11AM). The translation seems correct.
> 
> Because the profile is still active from yesterday. How do you solve this? If the schedule is 24h this will not work.

A daily profile ends at 00PM and restart at 00:00:01AM (or 00:00:59PM and restart at 00PM). A daily recurrency shall cover a 24h interval starting at 00PM. 

- When recurrencyKind is used in combination with a chargingSchedule duration shorter than the recurrencyKind period, the Charge Point SHALL fall back to default behaviour after the chargingSchedule duration ends. This fall back means that the Charge Point SHALL use a ChargingProfile with a lower stackLevel if available. If no other ChargingProfile is available, the Charge Point SHALL allow charging as if no ChargingProfile is installed. If the chargingSchedulePeriod and/or duration is longer then the recurrence period duration, the remainder periods SHALL not be executed.

 - When recurrencyKind is used in combination with a chargingSchedulePeriod and/or duration that is longer then the recurrence period duration, the remainder periods SHALL not be executed.

The code was already handling the first point. 
The merged code is not handling the second point. Discarding any period after the end of current day (or week) after the translation should not be difficult.

**Comment by OliveGerste** [comment]
> A daily profile ends at 00PM and restart at 00:00:01AM (or 00:00:59PM and restart at 00PM). A daily recurrency shall cover a 24h interval starting at 00PM

I did not find a reference in OCPP 1.6 spec, which says that a schedule starts and ends at 00:00. It just says:

- The schedule restarts every 24 hours, at the same time as in the startSchedule. for daily
- The schedule restarts every 7 days, at the same time and day-of-the-week as in the startSchedule. for weekly

**Comment by jerome-benoit** [comment]
> > A daily profile ends at 00PM and restart at 00:00:01AM (or 00:00:59PM and restart at 00PM). A daily recurrency shall cover a 24h interval starting at 00PM
> 
> I did not find a reference in OCPP 1.6 spec, which says that a schedule starts and ends at 00:00. It just says:
> 
> * The schedule restarts every 24 hours, at the same time as in the startSchedule. for daily
> * The schedule restarts every 7 days, at the same time and day-of-the-week as in the startSchedule. for weekly

Right, commit https://github.com/SAP/e-mobility-charging-stations-simulator/commit/522e4b058bc5ed1015cdf902e36a107490824f73 address the moving time windows. 
Thanks.

**Pull request merged**